### ISSUE TITLE
Register Vue Component as Renderer / Editor

### ISSUE THREAD

**Issue by aarondfrancis** [handsontable/vue-handsontable-official#129]
As far as I can tell from the docs, the only way to register a custom renderer or editor is to pass in a function or register it globally. Alternatively we can create `HotColumns` that serve as editors and renderers. What I would really like to do is register a vue component as a renderer or editor. Is that possible?

I see that internally, when `HotTable` sees a component is wraps it up to meet the function definition https://github.com/handsontable/vue-handsontable-official/blob/master/src/HotTable.vue#L129. Could that be exposed or shimmed somewhere?

My desired experience would be something like this: 

```js

// import the .vue component
import LinkRenderer from './renderers/link'; 

Handsontable.renderers.registerRenderer('LinkRenderer', LinkRenderer);
```

Is something like that possible now, or could it be added? 

Thanks!

**Comment by wojciechczerniak** [comment]
Unfortunately, no.

We've tried hard to expose `wrapComponent()` function to enable few more use cases. 
```js
import LinkRenderer from './renderers/link'; 
import { wrapComponent } fron '@handsontable/vue';
```

Mentioned `registerRenderer`:

```js
Handsontable.renderers.registerRenderer('LinkRenderer', wrapComponent(LinkRenderer));
```

Standard settings object:

```vue
<hot-table :settings={{ columns: [{ renderer: wrapComponent(LinkRenderer) }] }} />
```

And `<hot-column>` settings:

```vue
<hot-column :renderer={ wrapComponent(LinkRenderer) } />
```

But the code fragment you've already mentioned in the first line has a reference to `HotTable` instance:

https://github.com/handsontable/vue-handsontable-official/blob/6621371767ddf74b2224b54c85559551ed0595f3/src/HotTable.vue#L129-L132

We need it to cache components for virtual rendering and we have to keep them separately for each instance. Columns and rendereres may differ between instances. The main reason for cache is performance: we don't want to recreate Vue components on every Handsontable render.

**Comment by aarondfrancis** [comment]
What an incredibly thorough response, thank you! 

I clearly don't have the same grasp on the codebase as you so this may be naive, but would an optional third parameter to `getRedererWrapper` work? We're either allowed to pass in our our `LRUMap` or you use the one on the `$vm`.

**Comment by wojciechczerniak** [comment]
That might work. But would require some shared interface for LRUMap. I think we should start with gathering use cases first. Why the current implementation isn't working for you?

**Comment by aarondfrancis** [comment]
Sure, happy to elaborate.

I would like to have my template set up like this:

```vue
<hot-table :settings='hotSettings' />
```

but since I can't declare a Vue component as a renderer, I end up having to do something like this:

```vue

<hot-table :settings='hotSettings'>
    <template v-for='column in hotSettings.columns'>
        <hot-column v-if='column.type === "date"'>
            <date-renderer hot-renderer></date-renderer>
        </hot-column>

        <hot-column v-if='column.type === "link"'>
            <link-renderer hot-renderer></link-renderer>
        </hot-column>

        <hot-column v-if='column.type === "mention"'>
            <mention-renderer hot-renderer></mention-renderer>
        </hot-column>
    </template>
</hot-table>
```

You can see how quickly this explodes. With a couple of custom renderers and editors, it starts to get pretty messy and hard to reuse the functionality in a clean way. I suppose I could make my own `custom-hot-table` vue component that wraps all this up and kinda keeps it out of sight out of mind, but that's not my preference.

I have to say, handsontable is one of the most amazing pieces of OSS I've used, right up there with Vue and Laravel. It makes my life so easy. So please don't think I'm complaining! Just offering feedback.

**Comment by wojciechczerniak** [comment]
Thanks! Feedback is always welcomed. This is a valid use case, I just don't have a solution yet.

Maybe `@handsontable/vue` could have its own global register. But again, we cache components per instance. So we would need to rewrite it a little bit to read it from instance. Problem is that renderer has Handsontable instance but not HotTable. Or does it? 🤔 

For now, I wonder if that would work:

```vue
<hot-table :settings='hotSettings'>
      <hot-column v-for='column in hotSettings.columns' >
          <component hot-renderer v-is='column.renderer'></component>
      </hot-column>
</hot-table>
```

**Comment by wojciechczerniak** [comment]
Ok, I've got an idea.

We could read `HotTable` instance (and `rendererCache`) from Handsontable `instance`:
```js
getRendererWrapper: function (vNode: VNode): (...args) => HTMLElement {
        return function (instance, TD, row, col, prop, value, cellProperties) {
            const rendererCache = instance.HotTable.rendererCache;
```

This way, wrapped component has no hardcoded `$vm` instance. It would be possible register it globally like in the first comment: 
```js
Handsontable.renderers.registerRenderer('LinkRenderer', wrapComponent(LinkRenderer));
```
Handsontable then uses renderers as strings and handles the rest. I like that.

There are no additional arguments in the `wrapComponent` method. Nice.

We would need to link `HotTable` instance to Handsontable instance somewhere here:
https://github.com/handsontable/vue-handsontable-official/blob/6621371767ddf74b2224b54c85559551ed0595f3/src/HotTable.vue#L86-L87

But Handsontable would have to know about the wrapper, probably in `handsontable.d.ts`. I'm not sure if I like the idea that module inside knows about the wrapper module.

**Comment by aarondfrancis** [comment]
Your last suggestion seems awesome to me. Of course I can't speak to 
> But Handsontable would have to know about the wrapper, probably in handsontable.d.ts. I'm not sure if I like the idea that module inside knows about the wrapper module.

but should that prove workable, the solution seems great!

It's working just fine for now so think on it for a while, but this would be a nice improvement for me!

**Comment by h8art** [comment]
Any ideas how to register custom component as editor for specified cell (cellProperties.editor)?

**Comment by wojciechczerniak** [comment]
> Any ideas how to register custom component as editor for specified cell (cellProperties.editor)?

You can't. For the same reasons discussed above.

**Comment by mismith** [comment]
Has anyone else figured out how to do this? I couldn't get any of the snippets posted above to actually work, and am hoping something has changed since this was originally asked that would make this possible nowadays?

**Comment by pnowak** [comment]
Hi @mismith 

Unfortunately, we have not yet implemented the changes that @wojciechczerniak mentions above. 
And we don't have them on our roadmap yet.

**Comment by aarondfrancis** [comment]
@mismith As the complexity of what I needed grew, I ended up abandoning the vue wrapper altogether and just using the standard HoT library, taking advantage of custom renderers where I needed them. 

Clearly I'm not aware of your use case, but this has served me extremely well. I think the wrapper is quite nice for basic stuff though.

**Comment by mismith** [comment]
@aarondfrancis Yeah, I think I may just have to build the renderer is native DOM after all :(
I was hoping to be able to reuse some existing file upload logic (in Vue-land) instead, and was unsure whether the snippets above were just guesses/POC or whether they would actually function (albeit as hacks, presumably)
